### ISSUE TITLE
Random GCode Lines Skipped & Stopping in the middle of a Cycle.

### ISSUE THREAD

**Issue by antbarney** [synthetos/g2#209]
Well working on a Project over the holiday to make some engraved discs. I have found that Some times entire lines of Gcode will be skipped or the Cycle will just stop mid process. I am using ChiliPepper to stream the GCode and build 100.15 (Just did the Build this morning after pulling the latest updates). 

Here is one of the files I have had the most issues with. It will complete some times but most times it will stop in the middle of the cycle or miss sections in the middle. Just checking if this is related to an of the known issues being worked.

[Katelin_Front_0001.txt](https://github.com/synthetos/g2/files/673341/Katelin_Front_0001.txt)

**Comment by johnlauer** [comment]
My guess is cuz your lines and or decimals are too long.

Some of your gcode…

(Using default header. To add your own header create file "header" in the output dir.)

G02 X16.167383 Y19.391749 Z-1.000000 I-1.098619 J-0.120748 F300.000000

**Comment by antbarney** [comment]
I tried reducing that to 4 decimal places before but figured I would give it another go since you also suggested it. Also removed all Spaces and comments. Still have it stall from time to time. 

[Katelin_Test_4Digit_0002.txt](https://github.com/synthetos/g2/files/673416/Katelin_Test_4Digit_0002.txt)

**Comment by johnlauer** [comment]
Just ran it on tinyg v8. Ran like a champ.

When you say stall, what do you mean? Do you have to hit resume? Or does it just pause and resume on it’s own, cuz if so you’re not uploading lines fast enough.

**Comment by johnlauer** [comment]
Made a video of the perfect run

https://www.youtube.com/watch?v=3T4jllq-JZE

**Comment by antbarney** [comment]
First I should have mentioned this is on a Due running the latest g2core Edge Build with custom wiring running external drivers. 

When I say stall I mean the cycle stops with a number of lines in the SPJS buffer and hitting resume can some times get it to start again but that is normally only for the next line. I can recover if I wipe the buffer and restart, I think it is some new bug in the edge builds as I did not notice this before. Also removed my Limit switches to make sure it was not accidentally being triggered.

I have also tried now with decimal places down to 3 
[Katelin_Test_3Digit.txt](https://github.com/synthetos/g2/files/673445/Katelin_Test_3Digit.txt)


Screenshot of what ChiliPeppr Looks like after it hangs. 
![test](https://cloud.githubusercontent.com/assets/17607285/21488192/66b08d16-cb91-11e6-8a79-37644b01a705.png)

System info.
{"r":{"sys":{"fb":100.15,"fbs":"100.15-7-g9d372","fbc":"settings_cbeam.h","fv":0.99,"hp":3,"hv":0,"id":"0084-d639-18c6-9ce","jt":1.25,"ct":0.01,"sl":0,"lim":1,"saf":0,"m48e":1,"mfoe":0,"mfo":1,"mtoe":0,"mto":1,"mt":10,"spep":1,"spdp":0,"spph":1,"spdw":1,"ssoe":0,"sso":1,"cofp":1,"comp":1,"coph":1,"tv":1,"ej":1,"jv":2,"qv":0,"sv":1,"si":250,"gpl":0,"gun":1,"gco":1,"gpa":2,"gdi":0}},"f":[1,0,3]}

**Comment by johnlauer** [comment]
Notice in your sr responses it skips from line 78 to line 83. That looks whacky to me. It should have an sr for 79, 80, 81, 82, etc. without those sr’s, which is what spjs looks for to know the buffer flow can open back up again, you are dead in the water. When you hit resume you force the buffer back open until the same bug happens again. This is synthetos’s bug from what I can tell.

**Comment by antbarney** [comment]
Ya, I was pretty sure it was a bug in the TinyG2 Code. I also noticed that some of the response messages due contain errors after forcing the resume so I think some of the GCode is getting garbled/not fully buffered properly. As these sames lines will some times work.

some examples:
{"r":{},"f":[1,102,50]}
{"sr":{"stat":3}}
{"r":{},"f":[1,154,37]}

**Comment by giseburt** [comment]
The SR can skip lines. You can't rely on SR to report every line. If you
need to see every line be accepted, you have to set the JV to 4 so it'll
report the line number in n.

Explanation: SR will only report the current state when the report was
generated. Status reports are intentionally throttled and spaced out. They
are generally returned relatively evenly timed during a job, but if there
are a large amount of very quick moves (which requires extra planning time)
they may be throttled more than usual, and even more lines will be
"skipped."

All that said, I thought CP relied on the {r:...}, not the SR. Is this
correct?

-Rob

**Comment by johnlauer** [comment]
Oh, sorry, I meant TinyG buffer in spjs needs an r: response to know when to release the buffer, not an sr

**Comment by giseburt** [comment]
John: Ok. That's what I thought. But now anyone perusing this issue will
see why you can't rely on SR for each line. ;-)

Other than some fairly long lines I don't see anything that stands out as
an obvious problem. I'm on my phone now so I can't tell a lot. We'll look
into it and see what we can find.

-Rob

**Comment by antbarney** [comment]
Just to add some more info flashed back to the Master Branch build and ran this perfectly at least 3 times without any issues. Wanted to be sure to rule out anything in my build.

[fb]  firmware build             89.03
[fbs] firmware build "                           build"
[fv]  firmware version            0.98
[cv]  configuration version       7.00
[hp]  hardware platform           3.00
[hv]  hardware version            0.00
[id]  TinyG ID       0213-0215-d403

**Comment by aldenhart** [comment]
@artbarney -  Can you please post your settings_cbeam.h file?

**Comment by antbarney** [comment]
[settings_cbeam_edge.txt](https://github.com/synthetos/g2/files/678883/settings_cbeam_edge.txt)
[settings_cbeam_master.txt](https://github.com/synthetos/g2/files/678884/settings_cbeam_master.txt)

Here is the settings file I used to do the builds from both the edge and master branch

**Comment by stg** [comment]
If you are not running a mill/router, or you are getting this error while testing without the spindle motor running you can stop reading - otherwise I'd like to suggest EMC and static electricity as possible causes.

I've had this error many, many times early on. The controller would sometimes skip a random number of g-code lines and force the tool straight through the workpiece. Sometimes it would stop dead in its tracks. After a stop it could sometimes be recovered, sometimes not. The USB connection would sometimes break and require a reconnect before accepting anything at all. It all sounds too similar to not mention it - sorry for running my mouth if it does not apply, but here goes:

I was using a standard chinese spindle @1.5kW/24krpm. These motors put out an absurd amount of EMC especially when starting/stopping. More than enough to glitch out any nearby unprotected system and wreak havoc on any high impedance end-stops, touch-plate, other sensors. Cutting certain materials such as MDF and many plastics also cause build up of static electricity on the motor and Z axis as many machines are electrically insulated between Z and X as well as X and Y axes. You may notice static shocks when touching the motor during/after runs.

Both phenomena can enter and travel along cables to the controller - in my case also an Arduino Due which appears to be fairly sensitive.

To solve the problem I had to encapsulate the Arduino in a metal case and run external cables from a proper ground connection to the Dues new case, the CNC machine bed, and one point on each axis to make sure any static build up would discharge immediately. The last bit I needed for absolute stability was some ferrites on the cabling and a shorter, better shielded usb cable.

As for the sensors, one needs to make sure every signal in or near the machine is fairly low-impedance or heavily filtered - not true for out-of-the-box chinese machines.

Every time I accidentally disconnect ground (a gator clip on the nearest copper piped radiator) the problem comes back so by now I'd say it's 100% verified.

**Comment by antbarney** [comment]
Yes this is testing with the spindle off. And as stated only seems to affect the edge builds. Master branch builds look to work fine.

**Comment by cmaccioli** [comment]
I am having a similar issue. Just got the system set up last weekend and have been able to run small gcode samples, but when I try to run a large file, the systems hangs.  Sometimes it starts back up, but if I flush the planner buffer it starts running again.  It almost always stops when the planner hits 48. I just pulled the latest 100.17 and it is still happening. Running in chillipepper, with tinyg buffer (tried tinyg2 as well).  I can get it to run more lines by changing gcode sender to 100 pre-upload, no multiline and delay to 50 ms.  
This happens running a laser or spindle, I have ferrite beads on all low level signals, ground wire, etc.  Running on Due with no shield, 3.3v stepper to Intelligent motions IM804 stepper drivers.
Running JSON ver. 1.92 on windows 8 (x64).

**Comment by aldenhart** [comment]
@artbarney - we have not been able to make CP / 100.xx fail with the Kaitlin files. Do you have a longer file that does this? Also, what are you running as a host - Windows 10, OSX, rPi...?

@cmaccioli - can you post a file that fails for you?

**Comment by antbarney** [comment]
Sure Here is a bigger one I just tested, Fails at random spots also. All tests have been using ChiliPeppr SPJS 1.92, I have tried from Win10,rPi3,Ubuntu but all where with the SPJS 1.92
 
[Table.txt](https://github.com/synthetos/g2/files/681619/Table.txt)

**Comment by aldenhart** [comment]
@antbarney - Question for you - are you able to  run the default Gcode file that opens up with the CP TinyG workspace? I'm running this fine - before trying your files. If you cannot run it without issue it points to something in the toolchain - more than likely. I'm running on OSX 10.11, CP 1.92, set to 100 line pre-load. All other CP parameters are defaults.

**Comment by cmaccioli** [comment]
@aldenhart - Ill upload a file tonight when I get home.  I'll try the default Gcode (CP logo) when I get home as well. I'll also try some of the files posted by @antbarney. 

- I was using SPJS 1.88 with same issues, but updated yesterday to 1.92 to see if it made a difference (it did not).

Using Atmel Studio 7 to compile. Did have an error initially and it wouldn't compile, but had error in my settings file regarding ABSOLUTE_DISTANCE_MODE  not defined, but addressed that and compiled ok. 

Question: The planner status in CP, was typically at 48 when it stopped, does this mean the planner was overloaded or their was nothing planned?  After several tries, I left the machine stay stopped for 10-30 seconds and sometimes it would start back up and the planner would drop below 48.  
I was also able to continually press the ~ (cycl start) in CP and the machine would basically step through the program, sometimes running 1 line of code, sometimes running 2-3 lines.

Sorry of I am giving too much or impertinent info. I am new with tinyg2 and am learning as I go.  I am reading thru the code and wiki and gaining a better understanding, but it will take a while.

**Comment by johnlauer** [comment]
Sounds like same issue that about 3 other folks have mentioned. I think r:{}’s are disappearing coming back from TinyG and thus the buffer locks.

**Comment by antbarney** [comment]
I will try with the standard CP file when I get home today and report.

**Comment by ewidance** [comment]
I got exactly the same problem, using OSX / Firefox / SPJS 1.92 / Edge. I got no spindle, only a laser. I tried with many files, espacially one generated with arduino Uno 2 Top layer. It's stops very early in file. When i hit "%" (Queue flush) it starts again then stops.

**Comment by cmaccioli** [comment]
Was able to run the CP logo (mm) without issue.
Tried running my file and it stopped at various lines.
Tried running table.txt file from @antbarney and stopped at line 485.
Each time I tried to run it seems like it would stop earlier until I stopped JSON and restarted chrome.
[catcher_0006.txt](https://github.com/synthetos/g2/files/683534/catcher_0006.txt)

Also tried new USB cable just in case.
Could there be a driver issue? I know at first, I had to fudge the Tinyg.inf file because the ID of the port was different.

One time after it stopped, I turned off the filter in serial port console and hit the ~ button and here is what I got back : (it stopped at line1647, I hit ~ and it ran until line1664)
"~

{"qr":""}
{"Lbs":199}
{"r":{"n":1648},"f":[1,0,41]}
{"Lbs":183}
{"r":{"n":1649},"f":[1,0,27]}
{"qr":46}
{"Lbs":182}
{"r":{"n":1650},"f":[1,0,27]}
{"qr":45}
{"Lbs":165}
{"r":{"n":1651},"f":[1,0,43]}
{"Lbs":163}
{"r":{"n":1652},"f":[1,0,27]}
{"qr":43}
{"Lbs":136}
{"r":{"n":1653},"f":[1,0,12]}
{"qr":42}
{"Lbs":109}
{"r":{"n":1654},"f":[1,0,12]}
{"qr":40}
{"Lbs":66}
{"r":{"n":1655},"f":[1,0,15]}
{"Lbs":169}
{"r":{"n":1656},"f":[1,0,27]}
{"qr":39}
{"Lbs":157}
{"r":{"n":1657},"f":[1,0,12]}
{"qr":38}
{"Lbs":188}
{"r":{"n":1658},"f":[1,0,15]}
{"qr":36}
{"Lbs":173}
{"r":{"n":1659},"f":[1,0,12]}
{"qr":35}
{"Lbs":146}
{"r":{"n":1660},"f":[1,0,21]}
{"Lbs":134}
{"r":{"n":1661},"f":[1,0,43]}
{"Lbs":199}
{"r":{"n":1662},"f":[1,0,43]}
{"Lbs":187}
{"r":{"n":1663},"f":[1,0,37]}
{"Lbs":210}
{"r":{"n":1664},"f":[1,0,43]}
{"sr":{"line":1648,"posx":19.783,"posy":6.313,"vel":145.52,"stat":5,"momo":3,"mpox":214.783,"mpoy":181.313}}
{"qr":32}
{"sr":{"line":1651,"posx":19.971,"posy":6.597,"vel":50.48,"mpox":214.971,"mpoy":181.597}}
{"qr":35}
{"sr":{"line":1656,"posx":19.248,"posy":6.235,"vel":802.97,"feed":3000,"momo":1,"mpox":214.248,"mpoy":181.235}}
{"sr":{"posx":14.014,"posy":3.531,"vel":1563.36,"mpox":209.014,"mpoy":178.531}}
{"sr":{"posx":10.843,"posy":1.893,"vel":93.5,"mpox":205.843,"mpoy":176.893}}
{"sr":{"line":1662,"posx":11.027,"posy":2.17,"vel":198.59,"feed":500,"momo":3,"mpox":206.027,"mpoy":177.17}}
{"qr":46}
{"qr":48}
{"sr":{"line":1664,"posx":10.774,"posy":2.737,"vel":0,"stat":3,"mpox":205.774,"mpoy":177.737}}
{"sr":{"stat":3}}

**Comment by aldenhart** [comment]
@cmaccioli - Thanks for the detailed dump. I can't find a smoking gun, but here's what we can learn from it. 

- The "Lbs" commands are issued to SPJS only and are not sent to the g2 board. They can be ignored. (John, please correct me if I'm wrong about this).

- The "qr" commands are a response from g2 saying how many free planner buffers it has - up to a maximum of 48. If the planner is full this number is usually 3 (it keeps some free for odd commands). It returns 48 when the buffer is empty. 

- The fact that the buffers were hovering in the mid-levels (not full, at qr:3) indicates that there are very short Gcode lines being sent, and the buffer and play-out are throttling. g2 throttles the top speed of the Gcode lines if the sender / USB system cannot keep up with the rate the lines want to play out. The effect is a planner buffer that oscillates around the mid-levels as you see here. 

So what this tells us is that the buffer was almost empty at line 1648, then filled up some, then stopped receiving any new input after line 1664, for which it received a response ( {r:...} ). The {r} was returned, and a status report {sr:...} later, once it was executed. It looks like line 1664 was the last line sent t the board, as this coincides with the buffer emptying {qr:48}.

The question is - did SPJS ever send line 1665? If it did, is this logged, and was the {r} returned or not? from the look of this dump it appears that it was never sent, but I don't know the details of the logging.

Some potential diagnostics:

- Once stalled, query the g2 board with a {"line":null} command. It will return the last line number it received. See if this matches the dump. Was it 1664 or 1665 (or later)?

- Have the CP dump flag any non-zero status responses --> "f":[1,0,43] is right, "f":[1,100,43] is an example of an error code whose line cannot be parsed. ANY non-zero status response is an error, and should be flagged somehow. I don;t see any here.

**Comment by cmaccioli** [comment]
@aldenhart -
OK - for the buffer - The machine actually stopped at 1577 and I hit the ~ several times, then thought to turn off the filtering when it stopped at 1647 so I could see the status and saved the status info after that.
I am pretty sure that CP had sent all of the lines of code already.

Is there a way of outputting the status info to a text file directly while running?
Sorry if this is a stupid question, but how do I have CP dump flag any non-zero status responses?

I am connected directly to the USB port of my laptop, no Hub. 
Are we all seeing a USB port speed issue?
FYI, in windows device manager, the port shows up as "TinyG v2 (Data Channel) (Com 11)" would that be correct?
I will try to get more info now.

**Comment by cmaccioli** [comment]
So I tried running my file and it stopped at line 524, but CP kept sending info and the serial monitor, truncated the data.
Note that this is the earliest Line # I have seen it stop at.
Tried running the  {"line":null}, but got nothing in response.
I stopped GCODE, then flushed the planner and got the following:
%

{"qr":""}
{"er":{"fb":100.17,"st":157,"msg":"Arc specification error - missing offset(s) - arc offsets missing or zero"}}
{"Lbs":0}
{"r":{"n":525},"f":[1,157,26]}
{"r":{"qr":48},"f":[1,0,11]}
{"sr":{"line":525,"stat":2}}
  {"line":null}
{"Lbs":0}
{"r":{"line":525},"f":[1,0,15]}

**Comment by aldenhart** [comment]
The g2 board does not have memory for any logging. So all logging must be done at the host end, or on the USB wire. What I'd like to see from the log is:

1. Were there any non-zero status returns? Indicating either a corrupted transmission or some garbled transmission.

1. Were there any corrupted return lines that did not pass the JSON parser? Indicating a corrupted response. If the board is truncating a response line it won't pass JSON parsing (99.999% of the time!)

1. What was the last line number sent form SPJS, and what was the last line number received in an {r:}?

1. Did the board receive any lines where the line numbers did not monotonically increase by 1? Indicating a dropped line from SPJS.

Can you think of any other tests?

Be aware that for #3 the {r} test if the R "was not received" by the logger it does not necessarily mean that it was not sent by the board. That could be the case, or it could have been not picked up by the receiver, somehow. It's unlikely that the USB channel failed, as that would have most likely caused one of the corruption cases.

What we may need to do is get a USB sniffer to see what was actually sent in both directions. Wireshark can do this on OSX / Linux / Windows with a USBpcap so it might be necessary to configure a USB capture to really look under the covers.

I don't see the 100 builds completely missing a transmission - although this is possible - but it would not surprise me if the TX timing is somewhat different than the earlier releases.

**Comment by aldenhart** [comment]
To your comment - If there is an arc specification error - or any other Gcode error - the job will stop.

If it made it past this point before it indicates that something was supported, as not all the modal "requirements" for an arc were satisfied.

What is the Gcode for line 525 and the lines leading up to it?

**Comment by antbarney** [comment]
Is this a screw drive machine with multistart screws? - This is a Openbuilds Cbeam machine. ACME screws on all axis. 

Also, power at 1.0 is high. What stepper drivers are you driving?  Are you getting driver overheats? - I am using external TB6600 drivers and since I am only running Nema 23's at 2.8amp they are not even getting above Room temp. and with external drivers I don't think the Power setting even takes effect I could most likely just set those to 0 and be ok.

**Comment by antbarney** [comment]
I also just tried the standard Gcode file from Chilipeppr and it also failed. 

Host rPi3 SPJS 1.92 Presend 100 lines rest default.

Here is the info I got after I flushed the buffer. 

N279 G3 X-11.7011 Y4.72 I-2.2027 J-2.9617

 {"line":null}
%

{"qr":""}
{"er":{"fb":100.17,"st":157,"msg":"Arc specification error - missing offset(s) - arc offsets missing or zero"}}
{"r":{},"f":[1,157,30]}
{"r":{"qr":48},"f":[1,0,11]}
{"sr":{"line":279,"stat":2}}

**Comment by aldenhart** [comment]
Right. The power level is irrelevant for external drivers. You can turn it off.

Question - do the Toshiba drivers take an active LOW enable signal? Or did you invert the enable somehow?

**Comment by antbarney** [comment]
I did not do any inverting manually. So I think they are Active HIGH. 
This is the one I used specifically. 

https://www.amazon.com/gp/product/B016ZJS1FA/ref=oh_aui_search_detailpage?ie=UTF8&psc=1

**Comment by cmaccioli** [comment]
I am running THK Linear actuator (ball screw with guides) using NEMA 23 with IM804 drivers runnning 16 microsteps @ 3.2 amp.  Inverted enable signal.

Just ran again and ran the JSON server in verbose mode out to text file and am attaching it.
I preloaded 10,000 lines (full program = 2980 lines).

[log.txt](https://github.com/synthetos/g2/files/683713/log.txt)

**Comment by antbarney** [comment]
Also did a Verbose log of it failing with the standard CP logo file.

[Verbose_SPJS_Output.txt](https://github.com/synthetos/g2/files/683720/Verbose_SPJS_Output.txt)

**Comment by aldenhart** [comment]
Unfortunately the verbose SPJS log doesn't help me much. It has all manner of things mixed in an it's hard to find what the actual back and forth is masked. What I'm trying to get at is something like:

>>> Sent:  (stuff sent to board)
<<< Recd: (stuff received from board)

Ideally w/ timestamps. We've built this for some of our other integrations (e.g. Shopbot) and it's a good way to get the actual data exchanges.

On the v8 we tap the serial lines and feed them to a Logic protocol analyzer to see this data. But the g2 stuff is pure USB, so we don't have those signals.

**Comment by johnlauer** [comment]
It would be cool if somebody added a command line flag to SPJS like -log and then added 2 new lines of code in the buffer that ONLY wrote outgoing and incoming lines. It wouldn’t be super hard.

**Comment by aldenhart** [comment]
I just ran the CP Logo Gcode (mm) with no issues. Is this the file you are running that is failing?

**Comment by cmaccioli** [comment]
@aldenhart - I was able to run that file without issue.  Not sure which file @antbarney was running though.

**Comment by cmaccioli** [comment]
@aldenhart Downloading winshark. never used it, so are there any specific instructions to get the data you might need?

**Comment by giseburt** [comment]
This part looks interesting:

2017/01/04 01:35:02 serial.go:357: Consuming sendBuffered queue. ok:true,
d:N584 M5
, id:g584
2017/01/04 01:35:02 serial.go:357: Consuming sendBuffered queue. ok:true,
d:N585 M30
, id:g585
2017/01/04 01:35:02 serial.go:368: Hit default in select clause
2017/01/04 01:35:02 serial.go:371: Done consuming sendBuffered cmds. ctr:352
2017/01/04 01:35:02 bufferflow_tinygg2.go:754: Lock being released in TinyG
buffer
2017/01/04 01:35:02 bufferflow_tinygg2.go:867: Just sent release to b.sem
so we will not block the sending to serial port anymore.
2017/01/04 01:35:02 serial.go:383: itemsInBuffer:1
2017/01/04 01:35:02 bufferflow_tinygg2.go:246: Done blocking cuz got b.sem
semaphore release. ok:true, unblockType:2
2017/01/04 01:35:02 bufferflow_tinygg2.go:251: This was an unblock of type
2, which means we're being asked to wipe internal buffer. so return false.
2017/01/04 01:35:02 serialport.go:246: We got back from BlockUntilReady()
but apparently we must cancel this cmd
2017/01/04 01:35:02 serial.go:401: We need to unpause our internal buffer.

**Comment by cmaccioli** [comment]
Pretty sure his machine stopped a ways above that at 
2017/01/04 01:33:43 bufferflow_tinygg2.go:309: < {"sr":{"posx":-1.556,"posy":0.128,"vel":0,"stat":3,"mpox":-1.556,"mpoy":0.128,"cycs":0}}
2017/01/04 01:33:53 bufferflow_tinygg2.go:309: < {"sr":{"stat":3}}.

**Comment by antbarney** [comment]
Correct, the machine stopped before that, and I had to flush the buffer (send "%") before I could get {"line":null} to return.

in my log file I ran the CP Logo file from the Chilipeppr Page (mm version) I have ran in a few times and have had it complete successfully at least once. And failure points in the other cases where in different spots.

**Comment by aldenhart** [comment]
@antbarney - I know you said this before, but this thread is getting long. What machines are you running? This all sounds like comms problems on specific machines. I am having no troubles  running this from a clean OSX 10.11. (Otherwise we would not have pushed to edge yet.)

**Comment by antbarney** [comment]
As the Host for SPJS I am running it from a dedicated Raspberry Pi 3 with SPJS 1.92, but I have also tried from a Fresh Win 10 Machine, as well as Ubuntu 16.04.1. 

G2 is running on a Arduino Due with custom wiring for the drivers.

**Comment by aldenhart** [comment]
Did you get better results on W10 or Ubuntu? The rPi was known to be underpowered. I'm not sure about the 3's, though.

**Comment by antbarney** [comment]
No, had the same issues on Ubuntu and W10. The Pi 3 seems to handle it pretty well have been using it for a while without issue.

**Comment by aldenhart** [comment]
Do all the Kaitlins fail about the same? I think I should run them next. Until I can reproduce the issue I'm not sure what I can do. So far I've been working up to it and everything I've run completes without issues

**Comment by antbarney** [comment]
They never fail in the same spots from what I can tell but they do fail regularly. That's what seems odd to me is sometimes it will successfully complete, but it will fail pretty regularly for me anyway.

**Comment by giseburt** [comment]
In the log, starting at 2017/01/04 01:33:33 (I can't get my phone to copy
it) SPJS says it's pausing, then 4 {r...} responses go by. It never I
paused as far as I can tell. That looks like it's the root of the problem
in there.

**Comment by antbarney** [comment]
[CP_Logo_Fail.txt](https://github.com/synthetos/g2/files/683823/CP_Logo_Fail.txt)
I know you said that the SPJS logs where not much help but wanted to just add another set this time from a W10 Host. 

I have also removed all Limit switches from my setup just to make sure there is not some glitching going on there (which I have checked many times before this also and always confirmed via $in that the input states where correct.)

**Comment by antbarney** [comment]
I also just had it fail basically at the start of the CP Logo file. and looks like the response is an error 154 (STAT_ARC_HAS_IMPOSSIBLE_CENTER_POINT) which from what I can was for line N37 G2 F800.0 X31.7601 Y-2.7342 I-5.0775 J0.7343 which is valid if at the proper starting point of X31.8196 Y-2.4008 Z0 and will run fine when I manually execute. Could the g2 Buffer or planner be missing actions and not be in the correct location when some of the lines get executed causing it to enter an error state? Sorry I am not all to much help really a noob when it comes to stuff like this.


2017/01/03 18:44:41 bufferflow_tinygg2.go:309: < {"sr":{"line":16,"posx":38.563,"posy":-4.195,"posz":0,"vel":800,"feed":800,"momo":2,"mpox":38.563,"mpoy":-4.195,"mpoz":0}}
2017/01/03 18:44:42 bufferflow_tinygg2.go:309: < {"sr":{"line":20,"posx":35.934,"posy":-3.092,"mpox":35.928,"mpoy":-3.073}}
2017/01/03 18:44:42 bufferflow_tinygg2.go:309: < {"sr":{"line":23,"posx":36.155,"posy":0.141,"mpox":36.155,"mpoy":0.141}}
2017/01/03 18:44:42 hub.go:122: sendjson {"P":"COM7","Data":[{"D":"N38 G2 X31.2841 Y-3.7614 I-2.3626 J0.471\n","Id":"g38"},{"D":"N39 G2 X30.6633 Y-4.2692 I-1.5584 J1.2717\n","Id":"g39"},{"D":"N40 G2 X29.8824 Y-4.452 I-0.7659 J1.5128\n","Id":"g40"},{"D":"N41 G2 X29.1178 Y-4.2757 I-0.0154 J1.6797\n","Id":"g41"},{"D":"N42 G2 X28.5081 Y-3.7819 I0.9002 J1.7348\n","Id":"g42"},{"D":"N43 G2 X28.0487 Y-2.7942 I1.8061 J1.4408\n","Id":"g43"},{"D":"N44 G2 X27.9406 Y-1.7102 I4.813 J1.0272\n","Id":"g44"},{"D":"N45 G2 X28.0605 Y-0.5925 I4.765 J0.0543\n","Id":"g45"},{"D":"N46 G2 X28.5423 Y0.4231 I2.4581 J-0.5442\n","Id":"g46"},{"D":"N47 G2 X29.154 Y0.9543 I1.7404 J-1.3862\n","Id":"g47"}]}
2017/01/03 18:44:42 serial.go:252: Done with writeJson method
2017/01/03 18:44:42 serialport.go:233: Got p.sendBuffered. data:N38 G2 X31.2841 Y-3.7614 I-2.3626 J0.471\n, id:g38, pause:0
2017/01/03 18:44:42 bufferflow_tinygg2.go:192: BlockUntilReady(cmd:N38 G2 X31.2841 Y-3.7614 I-2.3626 J0.471
, id:g38) start
2017/01/03 18:44:42 bufferflow_tinygg2.go:225: New line length: 41, buffer size increased to:90
2017/01/03 18:44:42 serialport.go:233: Got p.sendBuffered. data:N39 G2 X30.6633 Y-4.2692 I-1.5584 J1.2717\n, id:g39, pause:0
2017/01/03 18:44:42 bufferflow_tinygg2.go:192: BlockUntilReady(cmd:N39 G2 X30.6633 Y-4.2692 I-1.5584 J1.2717
, id:g39) start
2017/01/03 18:44:42 bufferflow_tinygg2.go:225: New line length: 42, buffer size increased to:132
2017/01/03 18:44:42 bufferflow_tinygg2.go:232: It looks like the buffer is over the allowed size, so we are going to pause. Then when some incoming responses come in a check will occur to see if there's room to send this command. Pausing...
2017/01/03 18:44:42 bufferflow_tinygg2.go:243: Blocking on b.sem until told from OnIncomingData to go
2017/01/03 18:44:42 serialport.go:268: Got p.sendNoBuf. id:g38, pause:0, data:N38 G2 X31.2841 Y-3.7614 I-2.3626 J0.471\n
2017/01/03 18:44:42 serialport.go:275: Items In SPJS Queue List:9
2017/01/03 18:44:42 bufferflow_tinygg2.go:309: < {"r":{},"f":[1,154,41]}
2017/01/03 18:44:42 bufferflow_tinygg2.go:345: Buffer decreased to itemCnt:2, lenOfBuf:83
2017/01/03 18:44:42 bufferflow_tinygg2.go:867: Just sent release to b.sem so we will not block the sending to serial port anymore.
2017/01/03 18:44:42 bufferflow_tinygg2.go:246: Done blocking cuz got b.sem semaphore release. ok:true, unblockType:1
2017/01/03 18:44:42 serialport.go:233: Got p.sendBuffered. data:N40 G2 X29.8824 Y-4.452 I-0.7659 J1.5128\n, id:g40, pause:0
2017/01/03 18:44:42 bufferflow_tinygg2.go:192: BlockUntilReady(cmd:N40 G2 X29.8824 Y-4.452 I-0.7659 J1.5128
, id:g40) start
2017/01/03 18:44:42 bufferflow_tinygg2.go:225: New line length: 41, buffer size increased to:124
2017/01/03 18:44:42 bufferflow_tinygg2.go:232: It looks like the buffer is over the allowed size, so we are going to pause. Then when some incoming responses come in a check will occur to see if there's room to send this command. Pausing...
2017/01/03 18:44:42 bufferflow_tinygg2.go:243: Blocking on b.sem until told from OnIncomingData to go
2017/01/03 18:44:42 serialport.go:268: Got p.sendNoBuf. id:g39, pause:0, data:N39 G2 X30.6633 Y-4.2692 I-1.5584 J1.2717\n
2017/01/03 18:44:42 serialport.go:275: Items In SPJS Queue List:8
2017/01/03 18:44:42 bufferflow_tinygg2.go:309: < {"r":{},"f":[1,154,42]}
2017/01/03 18:44:42 bufferflow_tinygg2.go:345: Buffer decreased to itemCnt:2, lenOfBuf:83
2017/01/03 18:44:42 bufferflow_tinygg2.go:867: Just sent release to b.sem so we will not block the sending to serial port anymore.
2017/01/03 18:44:42 bufferflow_tinygg2.go:246: Done blocking cuz got b.sem semaphore release. ok:true, unblockType:1
2017/01/03 18:44:42 serialport.go:233: Got p.sendBuffered. data:N41 G2 X29.1178 Y-4.2757 I-0.0154 J1.6797\n, id:g41, pause:0
2017/01/03 18:44:42 bufferflow_tinygg2.go:192: BlockUntilReady(cmd:N41 G2 X29.1178 Y-4.2757 I-0.0154 J1.6797

**Comment by cmaccioli** [comment]
I have wireshark installed. any tips on how to run it? filters, etc.
I have a capture, but there is a ton of data, not sure how to properly filter it.

**Comment by cmaccioli** [comment]
Well, here is a wireshark log.  I filtered out interrupts and packets with data_len=0
[tinyg2_usb_log.zip](https://github.com/synthetos/g2/files/683872/tinyg2_usb_log.zip)
Zipped it, size was 274 mb

**Comment by ril3y** [comment]
That file looks like it has nothing in it?  Perhaps I am missing
something?  Wireshark just captures the encapsulated or encoded USB
packets?  I dont see an easy way to see the data inside these packets.
Again perhaps I am missing something.

**Comment by ril3y** [comment]
Nothing meaning nothing useful.  It has loads of USB packets.

**Comment by cmaccioli** [comment]
Just out of curiosity, it sounds like @antbarney and myself are compiling on our side. Just wondering if we gave you our settings file and any other changes, if you could compile on your side to see if it might be something in how we are compiling. Not sure if it is a stupid question or not, but grabbing at straws.

On a side note I have run the catcher0006.nc file over 10 times and it has never stopped at the same spot.

**Comment by cmaccioli** [comment]
@ril3y - I think you have to unzip and open with wireshark.
I had to filter out data that didn't look useful like packets with data length = 0 and packets that were interrupts. !(usb.transfer_type == 0x01) and !(usb.data_len == 0)
Then when I selected packets that had some data, I could see some info like"w	{"r":{"n":7},"f":[1,0,13]}"
![image](https://cloud.githubusercontent.com/assets/24707030/21631444/55a79f72-d209-11e6-873d-c8659f9aac95.png)

**Comment by ril3y** [comment]
Interesting. I did see that. I perhaps have an older version of Wireshark.
Will look again in the morning.

**Comment by cmaccioli** [comment]
Just in case, here is my settings file and the gshield pinout.
I had to change pwm pin as my pin 13 didn't seem to work properly.
pin_number kSpindle_PwmPinNumber    = 9;
pin_number kSpindle_Pwm2PinNumber   = -1;

[settings_noshield.h.txt](https://github.com/synthetos/g2/files/683909/settings_noshield.h.txt)
[gShield-pinout.h.txt](https://github.com/synthetos/g2/files/683910/gShield-pinout.h.txt)

**Comment by antbarney** [comment]
In case these end up being helpful in any way here is a Wireshark capture from my setup as well. Same filters as cmaccioli used this is doing the CP logo file.
[USB_Capture.zip](https://github.com/synthetos/g2/files/683955/USB_Capture.zip)

**Comment by antbarney** [comment]
I was digging into another Wireshark capture I took. Not sure if these are even helpful or not. But looked at the failure area.

This is one of my files Kaitlin files

Packets 85846 thru 163873
in this set we see a {"r":....} then we send N92 there is about a half second delay before we get our next {"r":...} in which case we send line N93. Then there is about 2 seconds before we get our next {"r":....} we send N94 and about a millisecond later we get {"r":...} so N95 is sent. Then no more responses after that.

[Failure.zip](https://github.com/synthetos/g2/files/684037/Failure.zip)

**Comment by cmaccioli** [comment]
Just an FYI, tried messing around last night with my laser and chillipeppr logo, and it failed.  I had previously run it 2-3 times.

**Comment by cmaccioli** [comment]
@antbarney - did you try running the circle diamond square program in CP. I was able to run it 2 times without issues.

**Comment by antbarney** [comment]
@cmaccioli  yes that is also successful for me ran it a few times without errors, I think the biggest difference is there is a limited number of Arc type movements in that file (mostly just straight lines).

**Comment by cmaccioli** [comment]
I was thinking the same thing, arc movements and/or really small movements.

**Comment by aldenhart** [comment]
@antbarney With regard to the delays in the {r} responses, communications do back up when arcs are being plotted, as new commands are not processed until the arc block (gcode line) is complete. I'm not sure it those timings are reasonable, as the arcs take less time than that to execute in my tests.

I'm still researching this one, but have not been able reproduce the failures testing under OSX. So I fired up CP and SPJS in a Windows 10 virtual machine. CP ran poorly under MS Edge, but perfectly under Chrome. CP in Edge could not keep up with the graphics, I guess. No failures on the logo or Kaitlin files.

If you have any pointers on how to reproduce the problem please let me know. Here's my toolchain:

- Macbook pro with OSX 10.11
- Parallels 12 VM with current Windows 10 (64 bit)
- CP workspace running on Google Chrome browser in W10
- SPJS 1.92 launched from Powershell in verbose mode
- USB connected through a high-quality hub
- TinyGv9

The only think I can see that's unusual is the v9. I will try this on a Due setup over the weekend. But I don't expect that will make a difference.

**Comment by antbarney** [comment]
@aldenhart 
Here is my setup for testing right now:

- Windows 10 Desktop

- SPJS 1.92 launched in verbose mode (I used cmd but I don't think that makes a difference)

- Directly connected to the USB

- CP Workspace is in Chrome on Windows 10

- Arduino Due with 100.18 build.

So yes the only thing different between my setup and yours from the sounds of it is the TinyGv9. And I also think the others who have comment on here are also using Due's

The other thing I am pretty sure I have mentioned is if I load up a Build from the Master branch everything works fine with these same files and setup.

**Comment by cmaccioli** [comment]
I'm running a Due, someone else referenced the issue above, not sure what they are running. @antbarney , did you load up a pre-compiled build from the master or did you compile it? What did you compile with?

**Comment by ewidance** [comment]
I'm also running a Due, with a custom shield, with same pinout as grblshield. Custom build omplied on OSX El Capitan. I've done tests using a grblshield and it's the same; 
Using Cambam, it's possible to customize post processor to use segments instead of arcs. I'll try it to see if it's related to arcs, or only to data volume.

**Comment by antbarney** [comment]
I compiled locally from an Ubuntu box for both the Edge build I have been testing with and the Master branch. (I just setup a separate local git repo for the master branch for testing)

**Comment by cmaccioli** [comment]
Not sure if this means anything or not.
I was browsing through the files in "g2core\board\ArduinoDue" and the other board directories (i.e. "g2core\board\G2v9"), and the only interesting thing I see is the motate_pin_assignments.h is quite different in the ArduinoDue directory than in all the other board directories.
ArduinoDue\motate_pin_assignments.h file uses -"_MAKE_MOTATE_PIN(0, 'A', 8);  // D0, RX0"
all other board\xxx\motate_pin_assignments.h file use "pin_number kSerial_RXPinNumber  = 0;"

**Comment by ewidance** [comment]
I just generated a gcode file with squares and circles, WITHOUT ANY ARC gcode command.
File is 26K and streams fine. I strongly suspect a problem with the Arc command (G2).

I attached the file for those who want to test...
[CamBamExport.ngc.txt](https://github.com/synthetos/g2/files/690511/CamBamExport.ngc.txt)

**Comment by aldenhart** [comment]
Thanks for the heads up. In looking over the SPJS logs that have been posted with the stall problem we've noticed that responses to arcs are delayed longer than they should be. However we have not seen this (not been able to reproduce this) on our setups yet. So this is where we stand. Updates as we get them.

**Comment by cmaccioli** [comment]
My uneducated (non-programmer) guess is how it is compiling or configured for the Due board. Everyone here with the issue is using the Due, Happens with different OS, browser, different compilers, ver 1.88 & 1.92 of JSON, buts runs fine on TinyGV9.

**Comment by antbarney** [comment]
For what its worth I also noticed some oddities today when messing around, when a stall occurs and I flush the buffer the response will normally show the last line executed, and in some cases that line number changed say it was executing line 3198 the line number in the response was 319 could the buffer be getting over full and dropping all of the stuff after the 8?

**Comment by cmaccioli** [comment]
I was able to run @ewidance file no prob. So to expand on the G2/G3 arc thing I ran a file that had no G2/G3 arc commands, just a ton of small G1 x,y moves (grayscale test for laser) and the machine hung up.  So either a ton of noise causing a stop from the small moves/stepper commands, processor overload or buffer issue? So being a non-programm and looking through the code the thing(s) that stand out are the "// TODO - FIX BUFFER OVERFLOW POTENTIAL" in  the gcode_parser.cpp  file and the _MAKE_MOTATE_PIN(0, 'A', 8); // D0, RX0" thing which is called from \board\ArduinoDue\motate_pin_assignments.h as well as in C:\Users\Hunter\Downloads\g2-edge\Motate\MotateProject\motate\board\ArduinoDue\motate_pin_assignments.h.
With that being said, gcode_parser.cpp seems to be used by all boards where as the motate_pin_assignments.h. thing only occurrs with the Due? Can it define the same Pin twice and cause issues?

**Comment by cmaccioli** [comment]
Sorry for all the post, just trying to give as much info as posssible.  I took @ewidance file, copied/pasted 3 times into 1 file and tried running it and it failed at line 2110 (original file was around 1500 lines).

**Comment by cmaccioli** [comment]
I downloaded pre-compiled version build 78.02 FV .97 and after reloading the tinyg.inf from master branch, i was able to run a program in full that I have yet been able to run.  Will try to compile master branch with my settings.

**Comment by ewidance** [comment]
The file i previously sent without G2 also hangs while sent repetively. G2 is definitely not the problem. 

An other idea is flow control management (COM_ENABLE_FLOW_CONTROL). FLOW_CONTROL_RTS don't seems to be usable on DUE, since RTS/CTS are not defined in pinout for Arduino DUE. That make sense, because ATMEGA8U2 serial converter only provides TX/RX, and there's no hardware handshake on embedded USB port. 

Those pins ARE defined and used in G2V9K (kSerial_RTSPinNumber / kSerial_CTSPinNumber ). I don't have any schematics for this board, but i suspect a serial converter chip WITH hardware handshake (FTDI?).

I checked my machine config, and i was setup for FLOW_CONTROL_RTS. That could be the source of problem. I'll try this afternoon with FLOW_CONTROL_XON, hoping SPJS server is able to manage that.

This could be a good explanation of char loss (lost in transmission ?).

**Comment by aldenhart** [comment]
Interesting observation. The fact that it degrades with multiple transmissions indicates that it may be something in the USB channel or SPJS that is "filling up". Can you try this with a soft RESET of g2core  between transmissions? Sending a control x (^x) from the serial console should do it - if printable ASCII gets through. I'm not sure how CP does reconnect, but I'd like to find a way to reset just the Due (g2core firmware) and not the SPJS connection if that's possible. The USB will of course reset, as it will have to re-enumerate when g2core comes back up. One thing that may help is to make the CP reconnect sequence irrelevant by burning in those parameters into the settings file. That way CP would not have to know it has a new connection. 

As for flow control, flow control is ignored when using USB, as USB provides its own flow control. The reason they are still in there is to support direct serial communication to g2 boards that support UART communications (which there are some of now).

**Comment by giseburt** [comment]
On the Due you cannot use the second USB connector (I believe it's labeled
as programming port), since the RTS line we use for flow control and
connection detection is wired to the reset pin of the processor.

As long as you use the native USB port it should work fine. There are
several UART and USARTs on that chip. You may be able to use another one,
but we haven't tested those configurations and guarantee that it'll work
without some code adjustments.

**Comment by ewidance** [comment]
Just tested with FLOW_CONTROL_XON on DUE. No changes. Gcode streaming arbritrary stops. Tried to push reset when stopped : no change. 

I'm sticked with this stupid problem...

**Comment by cmaccioli** [comment]
Maybe a foolish question, but what is different from the master that seems to allow  it to work? Is there more data being sent back/forth with the edge branch?

**Comment by ewidance** [comment]
@giseburt : I'm using a custom shield with a pinout closed to grblshield. Unfortunately all uarts pins are used by switches (x- /x+ ... z-/z+). Would be hard to test...

@cmaccioli Many differences.. Major rewrite and code reorganization...

**Comment by aldenhart** [comment]
Some progress to report. I was able to make it stall by using CP on OSX, SPJS on W10 in a VM, and a Due. Now on to do some HW debugging and find out if that last {r} was actually sent, or if it was still waiting to be sent. Two hypotheses to test (1) it was not sent - if so, why?, or (2) it was sent and somewhere in SPJS it was not picked up.

**Comment by ewidance** [comment]
@aldenhart @giseburt : Can you confirm edge is working fine on any gcode file, even for hours, or any board is affected ?

**Comment by aldenhart** [comment]
@ewidance We are running this exact code on other ARM boards with no issues. Multi-hour and overnight prints. It's this particular combination that has revealed the issue: OSX or W10 CP, SPJS 1.92 on W10, Due running 100.18 (or similar). 

That said, we found a smoking gun this morning - we are just not sure yet who's gun it is. There appears to be some buffer corruption in an incoming line of Gcode that causes a number of symptoms - one is SPJS getting out sync with {r}s, and another is that the line is not executed - so in many cases subsequent arcs will fail. 

Our next set of tests is to determine exactly where in the chain the corruption is occurring, then of course to fix it. 

We appreciate you patience and diligence. Let's nail this bug.

**Comment by amx1** [comment]
@aldenhart, @giseburt and all the others working on this,

it is the first time I am posting and therefore I would like to thank you all for your amazing work first.
Maybe I can also help a bit by providing some experience I made. I have some problems to use chilipeppr because I dont have a permanent internet connection in my workshop and thats why I played a bit with the CNC Software of Cheton Wu (https://github.com/cheton/cnc). The implementation of the TinyG2 protocol in chetons code is not working without problems and I tried to fix it.

The update to V100.17 changed the behaviour of the controller very much. It is now working much closer to the described protocol ( "r" handshake aka linemode), but I also saw the errors of dropping lines or even crashing the controller especially with ARC (G2 G3) or S.. or M commands also. You will find my modified subroutines of chetons code in the attachment. Take it as an  example for the implemented state engine. The problem for me was, that that the controller sends an {r: ..} always after  receiving the command independent of the status of the planner queue which leads to a buffer overflow if you dont check QR.  Simple commands and linear moves only generate one entry in the planner queue, but arcs and other commands could generate more entries automatically. Therefore checking the qr state before sending against a "low water mark" only is not sufficient. I implemented a wait state for the "complex" commands which stops the sender until the controller issues the next sr: line report. From my timing log, it seems that the controller sends the sr: line statement when starting the execution of the specific command. It has then enough space for the next arc or linear move, if the planner queue is not filling (qi == 0 or qo > qi). Up to now it is running pretty smooth with all my test cases. 

The challenge for me is, that the complex buffer chain and the variable reporting makes it very difficult to synch with the controller, if you want to fill the planner queue as much as possible. Chilipeppr was the benchmark for me. The "handshake" behaviour changed significantly from V100.16 to V100.17. The attached version only runs with V100.17 only.

An ideal solution for me would be, if the controller send the {r:..} as an acknowledge stating that it is ready to receive the next command independent of the type of the command and the status of the planner queue, because the simple handshake protocol: send a command wait until {r:..} and then send next command floods all buffers and crashes the controller even in V100.17.  

All test were made with Arduino due, V100.17, a custom grbl Shield and a CNC3040

Best

Axel  

 
[TinyG2_V100.17.zip](https://github.com/synthetos/g2/files/691778/TinyG2_V100.17.zip)

**Comment by johnlauer** [comment]
Why don’t you still just use Serial Port JSON Server but stream to it locally? That way you benefit from all the greatness of SPJS and wouldn’t have to reimplement a streaming protocol. You just connect to the websocket of SPJS on your local box.

**Comment by giseburt** [comment]
I think we're seeing a lot over over-complication on the sender side. A lot
of this is due to historical protocols getting mixed up with the current
protocol. Let me explain.

The simplest way to send to a g2core machine is to simply send lines. You
have to honor when to OS blocks, but that's usually just a setting when you
open the buffer. We'll call this the "blasting" technique.

The main down side of that "blasting data" technique is that the buffer on
the host side becomes large and you can't send a feed hold. Also, some
hosts (small Linux machines in particular) don't handle large send buffers
well and use more processor than they should (probably in paging the buffer
to/from RAM).

So, the next simplest way to send to a g2core machine, and the simplest
with a very responsive feed hold, is to:

1) send a line
2) wait for the {r} response
3) go to 1 if there are more lines to send
4) done

We'll call this "simple line mode."

There is one problem with that technique in that some moves are extremely
brief, and may take longer to transmit than to execute. So to avoid jerky
motion in this case we "prime" the channel with a few lines. Four seems to
work well.

0) Count how many lines you have to send (IOW, in the gcode file, etc) and
put that in lines_in_file. Set lines_to_send=4
1) while (lines_to_send && lines_in_file) { send_next_line();
lines_to_send--; lines_in_file--; }
2) read lines from serial, counting {r}s. For each {r}, set
lines_to_send++. (Error checking and status report processing goes here.)
3) if lines_in_file > 0 the goto (1)
4) done

It's best if 1 and 2 are run asynchronously, either in different threads or
event driven, but that's not strictly necessary.

We'll call this "complete line mode."

Notice that in none of these do we care about queue reports (which are
deprecated, btw), or do we care about characters sent, etc. I did gloss
over the error handling, which is parsing the lines coming back paying
attention to the {sr:{stat:...}}, error reports in {er:{...}}, and non-zero
status in the response footer {r:..., f:[...,status,...]}. Non-zero status
is generally non-fatal and can be ignored.

As you can see, you can't "blow out" the g2core receive buffer. It'll just
back up on the host side.

We deprecated queue reports (qr) just for what amx1 was saying about how
some gcode lines (like an arc) can fill an empty queue all by themselves.
There's no clear relationship between a "line" and a "queue entry," and it
was way too hard to keep track of them.

**Comment by amx1** [comment]
Thank you for your detailed explanation. The simple line mode works like charm with V100.17 now. The problem was the implementation of sending gcode commands.

The result of JSON.stringify ({ gc: gcode } ) result gc: and the gcode are doubled quoted and leads to an error 104, 

Simple test: (quotes are changed to down / up due to the mail program)

{„gc“: „N1 G1 X100 F100“}  leads to error 104

{ „gc“: N1 G1 X100 F100 } is ok

N1 G1 X100 F100 is ok too

I changed the sending routine to just sending the simple command string.

**Comment by giseburt** [comment]
{"gc":...} is for internal use. We probably should be throwing an error if
that comes in over the serial, since that will bypass the queue throttling
and eventually cause a crash.

**Comment by aldenhart** [comment]
Just pushed a fix to edge. The build and tag are now 100.19. 

The root case is errors that can occur on reception of a line of Gcode (host --> board) when running USB at 480 Mbps (High Speed).  The line is either dropped entirely, or garbled. The fix limits transmission to 12 Mbps.

This fix ran CP Logo and Kaitlin multiple times without issue with SPJS on a W10 (virtual machine). Please let us know your experiences. With enough feedback we'll close this issue - or keep at it.

We are still investigating. We don't know if this is a Windows10 issue as it appears there and not on other OSs (please let us know if you have experienced failures on other SPJS host OSs), or if the problem occurs in the USB drivers on the g2 board. It may be that the M3 core on the SAM3x can't handle comms at this rate. The reason the old code worked is that g2 never supported 480 mpbs, just 12 mpbs.

Also, the reason it appeared that arcs (G2/G3) were the problem is that if a line is missed then the arc will fail it's geometry tests as the starting point is different than what was expected. This issue is not a problem with arcs.

**Comment by antbarney** [comment]
Did some quick testing and so far this looks to have fixed it up. 

Tested the Kaitlin file as well at least 3 times without any missed lines or hangs.

CP running in Chrome on Ubuntu -> SPJS on a RPi3 -> TinyG2 (on a Due)

When I get some more time tomorrow I will also test from W10, but this config was only I had known failures with before as well.

**Comment by aldenhart** [comment]
Great. Thanks for checking.

**Comment by ewidance** [comment]
Quick test : Sent 29000 lines. Hangs at the end, Chilipeppr ended streaming, but remaining 2797 lines in transmission buffer are still there...

**Comment by ewidance** [comment]
Second try. Same results. Seems to hang when M30 reached in the serial port console, even if there's still chars to send in Serial Port JSON Server widget (here : 1303 bytes). Anyway, the good news is 28800 lines have been successfully transmitted. That far better than previously.

COnfig : Compilation on OSX, latest SPJS, Firefox on OSX, and Arduino DUE + custom shield.

**Comment by hurr1k4ne** [comment]
Let me hook up the debugging please,
i use a (compiled on linux, custom config) binary of the 108 edge branch.

And i don't have problems running code with lot of arcs and large program on windows 10 with Chilipeppr and SPJS 1.92. using a Lenovo Yoga with i5 8GB.
See Fusion 360 adaptive Clearing example:
![image](https://cloud.githubusercontent.com/assets/3234219/21751833/f6dd4212-d5cd-11e6-8c59-07df44e654e5.png)

But i have one occurence (don't know if it is corresponding):
When i put the machine with a full buffer (SPJS) in "shutdown" alarm state, i'm not able to send a $clear via the Chilipeppr Console.
If i disconnect (check box in CP) and reconnect (nothing else) i can send clear command through CP Console.
Is this intended / correct behaviour?

**Comment by ewidance** [comment]
Here's a screenshot of 28000 lines file that hangs.
![capture d ecran 2017-01-09 a 21 21 46](https://cloud.githubusercontent.com/assets/2135006/21781937/d9210008-d6b1-11e6-9597-d1214eb69ba5.png)

**Comment by giseburt** [comment]
@ewidance This appears to be an unrelated issue. You're seeing it stop on `M30`. `M30` is supposed to be an "end of job" line, and will indicate as much to Chilipeppr. It shouldn't cause it to stop sending lines, but that's a likely side-effect of the `M30`. However, that's not likely related to this issue directly.

@hurr1k4ne $clear is a text-mode command. We don't currently treat them as special as we do JSON commands, and it might not clear the alarm like the equivalent JSON (which is: `{clr:n}`). It should, but that's also not likely to be related to this particular issue.

**Comment by antbarney** [comment]
@giseburt @aldenhart 

Just wanted to add my additional testing info. Everything is looking good now and I have not been able to get it to miss any lines or get stuck after 10+ successful attempts from multiple configurations I have tried. Thanks so much for all the help on this one!

Most tests where from:
Chrome on Windows 10 -> SPJS on Win10 (ran locally on same machine) -> Due gShield Pinout.

**Comment by ewidance** [comment]
@giseburt   @aldenhart I've done some more tests. On firefox, hang just before the end as described. On Chrome, works.. but 2nd lauch does not starts. Buffer is filling, but machine does not start. I have to click tilde for remuming job.

**Comment by hurr1k4ne** [comment]
The Buffer problem might correspond to issue #225

**Comment by ewidance** [comment]
Yesterday night, made many tests on Chrome, with different gcode files up to 10 000 lines (DUE/SPJS on OSX/Gshiled pinout / gcc on osx) and worked except for the problem identified on #225. I'll test with fix for #225 during the day and will post a report. Seems to be on the right way!

**Comment by ewidance** [comment]
Can't make it works anymore, on Chrome. G2 stopped before the end, with 1000 to 3000 lines still in buffer. No idea of what changed...  I'm in 100.19 (DUE/SPJS on OSX/Gshiled pinout / gcc on osx)

**Comment by antbarney** [comment]
Was just going to also come comment, I just ran the entire Table.gcode file I uploaded here earlier. And it ran perfectly expect that it stopped after the M30 was sent. It will start up and finish by just sending a Cycle Start "~" and then it finished without issue. Is M30 bypassing the SPJS Buffer?

**Comment by chilipeppr** [comment]
Nothing bypasses the buffer but ! or % or ~

My gut is that M30 doesn't return an r:{} and thus SPJS thinks there's no room in TinyG buffer and it locks. When you send ~ to SPJS you effectively just tell it "i don't care what you think SPJS, just unlock the buffer and move along and send ~ to TinyG"

**Comment by giseburt** [comment]
M30 is treated as any other gcode and will return an {r:..} (
https://github.com/synthetos/g2/blob/edge/g2core/gcode_parser.cpp#L540 for
anyone who's curious), but the next status report will return a "stat":5
(program end) which otherwise wouldn't happen.

Technically, an M30 is supposed to be the last line of a file, and
indicates that the rest of the file is not to be read. See
https://www.tormach.com/m0_m1_m2_m30.html

So, I'm not sure if chili is right here or not. It IS ending the file, in a
manner of speaking, as the M30 requested. But it's a matter of
interpretation.

Why is there an M30 in the middle (or at least not at the very end) of a
file?

**Comment by antbarney** [comment]
Thatss the weird part there is not a M30 in the middle of the file. It is at the end of the file. What about M8? Mine Stopped and processed all lines in the buffer up until it switched to this comment (CIRCULAR1) Line 1351 in the file (1349 in the CP Gcode widget) 

[Table.txt](https://github.com/synthetos/g2/files/705716/Table.txt)

**Comment by hurr1k4ne** [comment]
Today i had a stop on a "arcy" Job. I ran 2/3 of the job and suddenly stopped.
g2core 100.19, windows 10, SPJS 1.92, Chrome (55?)

The adaptive clearing does contain a lot of arcs and it is the first time i got this happened...
I could do it a bit of movement when clicking feed hold and resume. But only when waiting a bit between...

i recorded a mobile phone video after it stopped and cleared the filter of the log...

maybe unusual, but maybe you find a clue...


[](https://youtu.be/iThBwtpTW0o)

**Comment by amx1** [comment]
I have the same problems even on a very simple implementation of the linemode protocol. (send command, wait for {"r:"...}, send next command) It works for at least most of the smaller cases ( < 2000 lines) very stable. Only for larger cases the transmission suddenly stops, the planner buffer will be processed normally until it is empty, the machine state changes to stop. In some cases I was able to continue the job by sending an "~" to resume. The problem is somewhat reproducable, but it is not predictable when the jobs stops, if it is after 10000 or 30000 lines. The main issue is that the controller doesnt send an {"r:".. } response in these cases

**Comment by amx1** [comment]
did some additional testing with my linemode sender implementation (I love this mode, because it is to simple to have bugs ;-))

Looking at the gcode showed that this is always for areas with a lot of very small movements due to approximation of curved areas (it happens for G1 and also for combination of G1, G2 and G3! The gcode is generated by Fusion 360, >65000 lines / pocket operations). I thought it might be a (planner) buffer problem, therefore changed the settings in planner.h by increasing (version 100.19, compiled on Ubuntu 16.04, TinyG2 on arduino due, custom grblshield, custom CNC3040 ballscrew ):

#define PLANNER_BUFFER_POOL_SIZE    (128)                // Suggest 12 min. Limit is 255
#define PLANNER_BUFFER_HEADROOM     (6)                 // Buffers to reserve in planner before processing new input line

This changed the behaviour, the jobs stops also, but a lot later - after 45000 - 50000 lines. So it is not really a planner buffer overflow. It might be a timing problem of the controller or the sender. It tested on a laptop directly connected and a PC directly connected) The last executed gcode in one case was a small linear move which lead to a segment time of 0,87 ms with the send feed rate (F1000).

My question is: what happens if the gcode segment leads to a smaller segment time than defined at
 
#define MIN_SEGMENT_MS              ((float)0.75)       // minimum segment milliseconds

Is the command ignored then and no {"r:"..} will be send? I also played a bit with MIN_SEGMENT_MS, but that is too much trail and error for me and I stopped it, because I don´t know the basics of the algorithm. The other issue could be dropping of some characters in the input buffer, because I could resume the job sometimes by sending !, and then ~. The controller missed the "non received" line then. I will do a basic test by throttling the sender, but needs some conding.

**Comment by ewidance** [comment]
@giseburt @aldenhart  Unfortunately, current version is totally unusable, since it's stops randomly before gcode streaming end. It seems current devs are more on 3D printer implementation than in core functions debugging. Lastest workaround (USB 1.1 instead of USB2) seems to be inoperant.

I have two machines wired with G2/Due/Custom Grbl shield: Once is a Proxxon KT150 mill, other is a corexy prototype (corexy.org) dedicated to laser engraving and cutting. I'm testing with Chili, maybe it's the problem? Is there an other way to stream a gcode file?

**Comment by cncgoko** [comment]
@ewidance I'm also testing using my own software (Goko, which works with TinyG) and I face the same issue. It randomly stops. For some reason, sometimes there is one missing {r:gc:{}}. So CP is not the cause here I think.
I'm working on G2Core support now and have issues with this as well on a Due with G2Core 100.17.

**Comment by amx1** [comment]
@ewidance, @giseburt @aldenhart 

I continued to play a bit with the different parameters and I can confirm the observations described in my previous post and also of cncgoko. It seems to be a buffer problem, the question was which one: the input buffer or the planner. I am on 100.19 and went back to the default values:

#define PLANNER_BUFFER_POOL_SIZE (48) // Suggest 12 min. Limit is 255
#define PLANNER_BUFFER_HEADROOM (4) // Buffers to reserve in planner before processing new input line

(changing these values changed the timing behaviour of the controller, but didn't solve the issue of sudden stopping the job) In most cases I can continue the job also by sending an ~. That means only the last \n is missing, but sometimes more characters were dropped. So this line or block is scrambled and has to be discarded. The error happens only when sending a lot of short moves. Typically the communication in these cases (sending command , waiting for {r:..., sending next command) is so fast that even positions are reported only every 10th to 15th line sent. The qr report is ok showing that the buffer is full qr=4, then an qo=1 -> {r:.. -> next command, qi=1 and so on.

I think it might be an input buffer problem or a problem with the USB communication

**Comment by giseburt** [comment]
Given that, as far as we can tell, this is only seen on a Due (please let us know if you're seeing it on a non-Due g2core machine), and that the 480Mbps to 12Mbps forced switch fixed the problem for some but not others, then that leads me to want to ask for one more test:

If you are still seeing this, are you plugging the Due into a USB3-capable port? If so, could you put a USB2 hub between the USB3 port and the Due, and see if that help?

As it stands, none of us on the dev team have been able to reproduce this since the 12Mbps lock. We have some new equipment coming in that will allow use to analyze the high-speed USB for signs of issues. Hopefully we can get this figured out, in spite of not being able to reproduce it "in the lab."

We appreciate your patience and continuing to provide feedback so we can get this resolved.

**Comment by hurr1k4ne** [comment]
Hey Rob,

Thank for the head-on this.

I had this problem happening in 100.19 plugged into a usb2 Port of my
Lenovo Yoga 13.(5y old) with a Anker USB cable.
Windows 10 chrome spjs1.92

**Comment by amx1** [comment]
Hi,

it makes no difference if it is directly connected to a dell laptop (USB2) or to a PC via USB2 hub (both on Ubuntu 16.04.1) for me too. I tested different optimisation levels, because I had some problems with different levels and interrupt routines in other projects in the past. O3 and O2 are even worse, but O0 doesn't compile:

Linking bin/sbv300-sbv300/g2core.elf 
Using linker script: ../Motate/MotateProject/motate/cmsis/TARGET_Atmel/linker_scripts/sam3x/sam3x8/gcc/flash.ld 
build/sbv300-sbv300/./temperature.o: In function `PID::getNewOutput(float)':
/home/am/workspace/git/TinyG2/g2core/./temperature.cpp:417: undefined reference to `PID::output_max'
collect2: error: ld returned 1 exit status
../Motate/MotateProject/motate/Motate.mk:446: recipe for target 'bin/sbv300-sbv300/g2core.elf' failed
make: *** [bin/sbv300-sbv300/g2core.elf] Error 1

but this shouldn't be the problem. I flashed also the version g2core-sbv300-sbv300-100.19.bin from the website, but I had the same error.

Please let me know if I can help you to debug or if I should check specific points. In all error cases we are missing the last one or two characters sent. I could continue the jobs by sending an ~.

**Comment by amx1** [comment]
Try to compile it with Option O1 instead of Os. I did not receive an error until now.

**Comment by aldenhart** [comment]
@ewidance - Can you please post the 28000 line file you are having issues with? Also your settings file (sorry if you have already done this - it's a long thread and I did not see it). Here's what I do have about your setup:
Config : Compilation on OSX, latest SPJS, Firefox on OSX, and Arduino DUE + custom shield.

**Comment by aldenhart** [comment]
Update: Merged a pull request that had a fix for some arc failures that could occur if an M30 followed an arc immediately. Probably not this issue, but it eliminates that case. 

Also, Rob has been spelunking some USB stuff using our shiny new analyzer and finding some things that appear to be affecting the W10 USB stack and possibly some of the *NIX stacks but not the OSX stack. So be on the lookout for another push to resolve these issues.

**Comment by hurr1k4ne** [comment]
I tried to mill a small SMD board with @aldenhart 's latest fix. 
I had two stops in 4 tries. 

Again: Win10, (native USB2 Port), SPJS 1.92, Due+gshieldv5, compiled on ubuntu linux with custom settings (derived from default). Firmware: 100.19 (b6995b28a6e7390e5a29d48b62796e239b8d4a21)

![image](https://cloud.githubusercontent.com/assets/3234219/22407870/3eb45fce-e66f-11e6-99b6-c76e538422c0.png)

**Comment by amx1** [comment]
@hurr1k4ne @aldenhart 

I played a bit with the compiler optimisation options. The default -Os is infact -O2 without the options increasing the size of the code. You can try to compile the Motate source with the option -fno-tree-ter which changes the behaviour of volatiles in interrupt routines. (I know that it is not really a systematic approach but ugly trial and error) The problem might be some sort of interrupt race condition. I had no stops since then. I can do some test runs with your gcode if you post it.

system; Ubuntu 16.04.1, 64 bit, SPJS 1.93, latest Firmware 100.19, compiled gcc 4.9.2 on Ubuntu, Test cases generated with Fusion 360 (postprocessor tinyg), 30000 lines, FreeCAD and HeeksCAD

What I did: I added the option -fno-tree-ter in file arm.mk (marked bold) 

path Motate/MotateProject/motate/arch

DEVICE_CPPFLAGS := -mcpu=$(CPU_DEV) -mthumb **-fno-tree-ter** -ffunction-sections -fdata-sections -std=gnu++14 -fno-rtti -fno-exceptions -fstrict-volatile-bitfields -fdiagnostics-show-option $(FLOAT_OPTIONS)

**Comment by hurr1k4ne** [comment]
Hey i just tried your fix, but it neither fixed the random stops nor auto-level shutdown (#232)

**Comment by aldenhart** [comment]
We have been going over the USB communications and how they affect various USB stacks. We will push shortly, and look for feedback. We think this should improve things a lot.

The autoleveler is a separate issue that is also being worked (#232).

**Comment by aldenhart** [comment]
Please check out Dev-209b and let us know your results. Ours test out well so far. Once we get confirmation on this we'll accept the PR into edge. This should take a few days, depending on the level of testing.

**Comment by amx1** [comment]
I did a clean check out  of dev-209b-random-pauses, everything is fine for me up to now. Did a check with a 65k line Fusion360 milling job (100 min) and other test files. no stops at all. environment: serial-json 1.93 running on Ubuntu 16.04, 64bit,  X86 PC, front end chrome on Linux and chrome on windows 10 Laptops via WLAN

@aldenhart @giseburt thank you for this great software!

**Comment by johnlauer** [comment]
That’s really good news. Maybe this issue is licked then!

**Comment by aldenhart** [comment]
Thanks for the testing. We continue to run this though it's paces and have found some other edge cases - mostly with really high speed jobs. Expect some further updates.

**Comment by amx1** [comment]
I am also working on a simple sender routine based on the linemode protocol for a different program, just by sending the commands and waiting for the {r: n: response and then sending next command. This simple handshake works without crashing or stops, but there is one effect with short lines (maybe the mentioned high speed jobs??). I did not measure the time exactly, but if the moves are to fast (processing time of the command in the order of the $si set time, default 250ms). the send loop is executed without any response of the controller except the {r:. as ACK. It is a bit like a DOS attack of the sender. The controller executes the commands, but there is no position information send back and the control program will not know the actual coordinates then. I implemented a work around by checking the qr status. If qr is <= 4 the sender is stopped until the value is >= 7. This introduces some delay and the controller will send the status changes including the new position data without the planner queue running out of commands. I know it is a hack, but it solves my problem.

**Comment by hurr1k4ne** [comment]
I did some quick testing too and had no stop with two different PCB Files.
Those had These Problem very often.

But you are right, this was Not a "high speed" Job.

Ubuntu 16.10, SPJS 1.94, Chromium, Chilipeppr, 100.19-dev-209b

**Comment by aldenhart** [comment]
I can deal wth hacks. I think what you are experiencing is that status reports are dialed back as low priority of the processor is having trouble keeping up with receiving and planning new moves. 

This was put in pace when the comm channel was running at 115,200 or 1 Mbps. It may be time to revisit that if we know we have a higher speed channel. This is something to keep an eye on. In the mean time your hack is a good one. It should find those cases where the throttling occurs.

**Comment by hurr1k4ne** [comment]
O i See amx1 mentioning the Sender Position Report. I faced this Problem
too.
I Set machine Zero and entered a g0 y100. Machine started moving, but at
about 50 the Chilipeppr Position info Stopped. After moving 1 Millimeter
the Position was correct again.

**Comment by giseburt** [comment]
amx1:

You are seeing when the system is below "Phat City Time". It's set in
planner.h (https://github.com/synthetos/g2/blob/edge/g2core/planner.h#L257)
and means that if the planner is not empty and has moves that, if executed
optimally (not ramping), total up to less than Phat City Time ms, then stop
status reports until we are out of that state.

The problem, as you said, was of starvation. It is rather expensive to send
a SR. If we do it when we really need to plan as fast as possible then
mostion will slow down.

If you would rather the slowdown than the SR delay, then lower Phat City (I
believe zero would be Bad.)

-Rob

**Comment by amx1** [comment]
I saw the Phat City Time parameter, but I decided it is not worth playing with it and detuning the whole controller. It can perfecly live with the high / low marks, because - as you said - it throttles the sender when needed. I don't need the exact real time position, it is just for the GUI. The GUI stutters sometime, when sending is even faster than the high/low mark as seen by hurr1k4ne, but doesn't execution of the commands. The planner queue was always filled sufficiently with my test cases.

**Comment by amx1** [comment]
@giseburt @aldenhart
everything runs stable, no crashes, no blocking until now, but there is one thing I saw during the tests: The feed hold command "!" doesn´t stop the motion immediatly. I tested a simple milling job generated with fusion360 (see zip file). Most of the operation is defined by arcs. If you try to stop the motion sending a feed hold "!" during the long arc operations, the controller stops only after completion of the arc command. It stops after a up to 180 degree turn. In case of intervention  you can´t stop the controller by feed hold.

[Lampenhalterung_Oben_1003.zip](https://github.com/synthetos/g2/files/752875/Lampenhalterung_Oben_1003.zip)

Env: sjs 1.93 on Ubuntu 16.04, frontend Win8.1, fusion 360, chrome browser, same result on ubuntu 16.04, cnc 1.9.0 alpha / cncjs by  cheton wu

**Comment by amx1** [comment]
@giseburt @aldenhart 

I did some further testing with different arc codes, same test case generated with different options of the tinyg.cps Fusion 360 postprocessor. You will find the test cases in the attached zip.

[Test_zirkular.zip](https://github.com/synthetos/g2/files/752929/Test_zirkular.zip)

1) Test_zirkular_2D_IJK.gcode / Option useRadius=no
no error 

2) Test_zirkular_2D_radius / option Option useRadius=yes
{"er":{"fb":100.19,"st":157,"msg":"Arc specification error - missing offset(s) - arc offsets missing or zero"}}

3) Test_zirkular_3D-contour_IJK.gcode / helical milling with F1200 for x,y,z / Option useRadius=no
x,y only operation ok, vel = 1200  but vel << 1200 in regions of x,y,z motion

4) Test_zirkular_3D-contour-radius.gcode / same as Test_zirkular_3D-contour_IJK.gcode but Option useRadius=yes

command: N22 G3 X16.626 Y42.381 R0.3 leads to error:
{"er":{"fb":100.19,"st":157,"msg":"Arc specification error - missing offset(s) - arc offsets missing or zero"}}

Test files: 
[Test_zirkular.zip](https://github.com/synthetos/g2/files/752936/Test_zirkular.zip)

**Comment by giseburt** [comment]
As far as we can tell, this issue has been fixed for a while now (with #236 being closed). Please let us know if this is wrong.

**Comment by jamescrypto** [comment]
Hi im having a problem with cncjs when i start Grbl Gcode the cnc starts and moves to where its about to start cutting then will just stop. if i try to resume it will do nothing. can any one help me i have attached a video showing what is happening and a file of the code processed from fusion360. i had to delete some of the comments and the file comments to get the cncjs to even start. but that is all i can get so ill upload the original processed file to. thank you for the help


https://user-images.githubusercontent.com/34698113/104114790-3ff13100-52bd-11eb-9e85-5bc3613726e9.mp4



[NEW 2001.txt](https://github.com/synthetos/g2/files/5792258/NEW.2001.txt)

**Comment by MitchBradley** [comment]
Why are you asking a GRBL question on a g2core forum?  GRBL and g2core are different.  In any case, your GCode contains G43 commands which neither GRBL nor g2core support.  You need to use the right postprocessor for whatever controller you are using. Fusion's grbl post is good for grbl, and its tinyg post is good for g2core.
